File: sinain-mobile-custom/ios/ISinain/Pipeline/FrameAnalyzer.swift
import Foundation
import UIKit
import Accelerate
import Vision

/// Analyzes JPEG frames for blur, brightness, perceptual hash, and text regions.
/// Runs entirely on a dedicated background queue — never blocks the main thread.
final class FrameAnalyzer: FrameAnalyzing {

    private let log = PipelineLogger(subsystem: "FrameAnalyzer")
    private let analysisQueue = DispatchQueue(label: "com.isinain.analysis", qos: .userInitiated)

    func analyze(_ jpegData: Data) async -> FrameAnalysis? {
        await withCheckedContinuation { continuation in
            analysisQueue.async {
                let t0 = CFAbsoluteTimeGetCurrent()

                guard let uiImage = UIImage(data: jpegData),
                      let cgImage = uiImage.cgImage else {
                    continuation.resume(returning: nil)
                    return
                }

                let width = cgImage.width
                let height = cgImage.height

                // 1. Grayscale conversion
                let grayscale = Self.cgImageToGrayscale(cgImage)

                // 2. Blur score (Laplacian variance)
                let blurScore = Self.computeBlurScore(grayscale, width: width, height: height)

                // 3. Brightness (mean)
                let brightnessAvg = Self.computeBrightness(grayscale)

                // 4. Perceptual dHash
                let perceptualHash = Self.computeDHash(cgImage)

                // 5. Text regions via Vision framework (async continuation, NOT semaphore)
                Self.detectTextRegions(cgImage: cgImage) { regions in
                    let analysisMs = (CFAbsoluteTimeGetCurrent() - t0) * 1000.0

                    let textRegionCount = regions.count
                    let textRegionConfidences = regions.map { $0.confidence }
                    let sortedTexts = regions
                        .sorted { $0.y < $1.y }
                        .map { $0.text }
                        .filter { !$0.isEmpty }
                    let nativeOcrText = sortedTexts.joined(separator: "\n")

                    let result = FrameAnalysis(
                        blurScore: blurScore,
                        brightnessAvg: brightnessAvg,
                        perceptualHash: perceptualHash,
                        textRegionCount: textRegionCount,
                        textRegionConfidences: textRegionConfidences,
                        nativeOcrText: nativeOcrText,
                        analysisMs: round(analysisMs * 10) / 10
                    )
                    continuation.resume(returning: result)
                }
            }
        }
    }

    // MARK: - Analysis Helpers

    /// Convert CGImage to grayscale pixel buffer.
    private static func cgImageToGrayscale(_ image: CGImage) -> [UInt8] {
        let width = image.width
        let height = image.height
        var pixels = [UInt8](repeating: 0, count: width * height)

        let colorSpace = CGColorSpaceCreateDeviceGray()
        guard let context = CGContext(
            data: &pixels,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: width,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.none.rawValue
        ) else { return pixels }

        context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
        return pixels
    }

    /// Blur score via Laplacian kernel variance (Accelerate framework).
    private static func computeBlurScore(_ grayscale: [UInt8], width: Int, height: Int) -> Double {
        let kernel: [Int16] = [0, 1, 0, 1, -4, 1, 0, 1, 0]

        var src = vImage_Buffer(
            data: UnsafeMutablePointer(mutating: grayscale),
            height: vImagePixelCount(height),
            width: vImagePixelCount(width),
            rowBytes: width
        )

        var destPixels = [UInt8](repeating: 0, count: width * height)
        var dest = vImage_Buffer(
            data: &destPixels,
            height: vImagePixelCount(height),
            width: vImagePixelCount(width),
            rowBytes: width
        )

        kernel.withUnsafeBufferPointer { kernelPtr in
            vImageConvolve_Planar8(
                &src, &dest, nil,
                0, 0,
                kernelPtr.baseAddress!, 3, 3,
                0,
                128,
                vImage_Flags(kvImageEdgeExtend)
            )
        }

        var floats = [Float](repeating: 0, count: width * height)
        vDSP.convertElements(of: destPixels, to: &floats)

        var mean: Float = 0
        var meanSq: Float = 0
        vDSP_meanv(floats, 1, &mean, vDSP_Length(floats.count))

        var squared = [Float](repeating: 0, count: floats.count)
        vDSP_vsq(floats, 1, &squared, 1, vDSP_Length(floats.count))
        vDSP_meanv(squared, 1, &meanSq, vDSP_Length(floats.count))

        return Double(meanSq - mean * mean)
    }

    /// Mean brightness of grayscale pixels.
    private static func computeBrightness(_ grayscale: [UInt8]) -> Double {
        var floats = [Float](repeating: 0, count: grayscale.count)
        vDSP.convertElements(of: grayscale, to: &floats)
        var mean: Float = 0
        vDSP_meanv(floats, 1, &mean, vDSP_Length(floats.count))
        return Double(mean)
    }

    /// Perceptual dHash: downscale to 9x8, compute horizontal gradient, 64-bit hash.
    private static func computeDHash(_ image: CGImage) -> String {
        let dw = 9, dh = 8
        var pixels = [UInt8](repeating: 0, count: dw * dh)

        let colorSpace = CGColorSpaceCreateDeviceGray()
        guard let ctx = CGContext(
            data: &pixels,
            width: dw,
            height: dh,
            bitsPerComponent: 8,
            bytesPerRow: dw,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.none.rawValue
        ) else { return "0000000000000000" }

        ctx.interpolationQuality = .medium
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: dw, height: dh))

        var hash: UInt64 = 0
        var bit = 0
        for row in 0..<dh {
            for col in 0..<(dw - 1) {
                if pixels[row * dw + col] > pixels[row * dw + col + 1] {
                    hash |= (1 << bit)
                }
                bit += 1
            }
        }

        return String(format: "%016llx", hash)
    }

    /// Text region detection via Vision framework (.accurate recognition level).
    /// Uses a completion handler — NOT a semaphore.
    private static func detectTextRegions(
        cgImage: CGImage,
        completion: @escaping ([(x: Double, y: Double, width: Double, height: Double, confidence: Double, text: String)]) -> Void
    ) {
        let request = VNRecognizeTextRequest { request, error in
            guard error == nil,
                  let observations = request.results as? [VNRecognizedTextObservation] else {
                completion([])
                return
            }

            let regions = observations.map { obs in
                let box = obs.boundingBox
                let text = obs.topCandidates(1).first?.string ?? ""
                return (
                    x: Double(box.origin.x),
                    y: Double(1.0 - box.origin.y - box.size.height),
                    width: Double(box.size.width),
                    height: Double(box.size.height),
                    confidence: Double(obs.confidence),
                    text: text
                )
            }
            completion(regions)
        }

        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true

        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        do {
            try handler.perform([request])
        } catch {
            completion([])
        }
    }
}

File: sinain-mobile/android/app/src/main/java/com/isinain/pipeline/FrameAnalyzer.kt
package com.isinain.pipeline

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume
import kotlin.math.abs

/**
 * Analyzes JPEG frames for blur, brightness, perceptual hash, and text regions.
 *
 * Android equivalent of iOS FrameAnalyzer.swift:
 *   - Accelerate/vImage → manual pixel operations on Bitmap
 *   - Vision framework → ML Kit Text Recognition
 *   - CGContext grayscale → Bitmap + ColorMatrix desaturation
 */
class FrameAnalyzer : FrameAnalyzing {

    private val log = PipelineLogger("FrameAnalyzer")
    private val textRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

    override suspend fun analyze(jpegData: ByteArray): FrameAnalysis? = withContext(Dispatchers.Default) {
        val t0 = System.nanoTime()

        val bitmap = BitmapFactory.decodeByteArray(jpegData, 0, jpegData.size) ?: run {
            log.warn("failed to decode JPEG")
            return@withContext null
        }

        val width = bitmap.width
        val height = bitmap.height

        // 1. Grayscale conversion
        val grayscale = bitmapToGrayscale(bitmap)

        // 2. Blur score (Laplacian variance)
        val blurScore = computeBlurScore(grayscale, width, height)

        // 3. Brightness (mean)
        val brightnessAvg = computeBrightness(grayscale)

        // 4. Perceptual dHash
        val perceptualHash = computeDHash(bitmap)

        // 5. Text regions via ML Kit
        val textResult = detectTextRegions(bitmap)

        bitmap.recycle()

        val analysisMs = (System.nanoTime() - t0) / 1_000_000.0

        FrameAnalysis(
            blurScore = blurScore,
            brightnessAvg = brightnessAvg,
            perceptualHash = perceptualHash,
            textRegionCount = textResult.regionCount,
            textRegionConfidences = textResult.confidences,
            nativeOcrText = textResult.text,
            analysisMs = (analysisMs * 10).toLong() / 10.0
        )
    }

    // ── Grayscale conversion ──────────────────────────────────

    private fun bitmapToGrayscale(bitmap: Bitmap): IntArray {
        val w = bitmap.width
        val h = bitmap.height
        val pixels = IntArray(w * h)
        bitmap.getPixels(pixels, 0, w, 0, 0, w, h)

        // Convert ARGB to grayscale using luminance formula
        val gray = IntArray(w * h)
        for (i in pixels.indices) {
            val p = pixels[i]
            val r = (p shr 16) and 0xFF
            val g = (p shr 8) and 0xFF
            val b = p and 0xFF
            // ITU-R BT.601 luminance
            gray[i] = (0.299 * r + 0.587 * g + 0.114 * b).toInt()
        }
        return gray
    }

    // ── Blur detection (Laplacian variance) ───────────────────

    /**
     * Applies 3x3 Laplacian kernel [0,1,0,1,-4,1,0,1,0] then computes variance.
     * Higher variance = sharper image.
     * Equivalent to iOS vImageConvolve_Planar8 + vDSP variance.
     */
    private fun computeBlurScore(gray: IntArray, width: Int, height: Int): Double {
        if (width < 3 || height < 3) return 0.0

        var sum = 0.0
        var sumSq = 0.0
        var count = 0

        for (y in 1 until height - 1) {
            for (x in 1 until width - 1) {
                val idx = y * width + x
                // Laplacian: center * -4 + top + bottom + left + right
                val laplacian = (
                    gray[idx - width] +        // top
                    gray[idx + width] +        // bottom
                    gray[idx - 1] +            // left
                    gray[idx + 1] +            // right
                    gray[idx] * -4             // center
                )
                val v = laplacian + 128        // bias to match iOS (unsigned output)
                sum += v
                sumSq += v.toDouble() * v
                count++
            }
        }

        if (count == 0) return 0.0
        val mean = sum / count
        val meanSq = sumSq / count
        return meanSq - mean * mean
    }

    // ── Brightness (mean of grayscale) ────────────────────────

    private fun computeBrightness(gray: IntArray): Double {
        if (gray.isEmpty()) return 0.0
        var sum = 0L
        for (v in gray) sum += v
        return sum.toDouble() / gray.size
    }

    // ── Perceptual dHash ──────────────────────────────────────

    /**
     * Downscale to 9x8 grayscale, compute horizontal gradient, 64-bit hash.
     * Identical algorithm to iOS computeDHash.
     */
    private fun computeDHash(bitmap: Bitmap): String {
        val dw = 9
        val dh = 8

        // Scale down to 9x8
        val scaled = Bitmap.createScaledBitmap(bitmap, dw, dh, true)
        val grayScaled = bitmapToGrayscale(scaled)
        if (scaled !== bitmap) scaled.recycle()

        var hash = 0L
        var bit = 0
        for (row in 0 until dh) {
            for (col in 0 until dw - 1) {
                if (grayScaled[row * dw + col] > grayScaled[row * dw + col + 1]) {
                    hash = hash or (1L shl bit)
                }
                bit++
            }
        }

        return String.format("%016x", hash)
    }

    // ── Text detection (ML Kit) ───────────────────────────────

    private data class TextResult(
        val regionCount: Int,
        val confidences: List<Double>,
        val text: String
    )

    /**
     * Uses ML Kit Text Recognition v2 — equivalent to iOS VNRecognizeTextRequest.
     * Returns recognized text blocks sorted by vertical position.
     */
    private suspend fun detectTextRegions(bitmap: Bitmap): TextResult {
        return suspendCancellableCoroutine { cont ->
            val inputImage = InputImage.fromBitmap(bitmap, 0)

            textRecognizer.process(inputImage)
                .addOnSuccessListener { visionText ->
                    val blocks = visionText.textBlocks
                    val regionCount = blocks.size

                    // ML Kit doesn't expose per-block confidence directly,
                    // but lines have confidence on Android API 26+.
                    // Use block-level heuristic: if text was recognized, confidence ≈ 0.8
                    val confidences = blocks.map { block ->
                        // Use the average line confidence if available,
                        // otherwise assume 0.8 for any recognized block
                        val lineConfs = block.lines.mapNotNull { line ->
                            line.confidence?.toDouble()
                        }
                        if (lineConfs.isNotEmpty()) lineConfs.average() else 0.8
                    }

                    // Sort blocks by vertical position (top of bounding box)
                    val sortedTexts = blocks
                        .sortedBy { it.boundingBox?.top ?: 0 }
                        .map { it.text }

                    val ocrText = sortedTexts.joinToString("\n")

                    cont.resume(TextResult(regionCount, confidences, ocrText))
                }
                .addOnFailureListener { e ->
                    log.warn("ML Kit text recognition failed: ${e.message}")
                    cont.resume(TextResult(0, emptyList(), ""))
                }
        }
    }
}
